MSBuild cannot access environment variables (PATH) for NuGet dependencies

### Tested versions

- Reproducible in: v4.3.stable.mono.official [77dcf97d8]

### System information

Godot v4.3.stable.mono - macOS 14.4.1 - Vulkan (Forward+) - integrated Apple M2 Pro - Apple M2 Pro (12 Threads)

### Issue description

My project has a dependency that is installed with NuGet. When I build the Godot project, this NuGet package tries to build an MSBuild target, and fails, because it cannot find `dotnet`, although `dotnet` is installed properly on my system.

### Steps to reproduce

I have created an example project with a dependency, added to the `.csproj` file:

    <PackageReference Include="FlatSharp.Compiler" Version="[7.7.0]" />

When I build the Godot project, this package tries to build an MSBuild target, and fails, because it cannot find `dotnet`. This is the output in Godot's "MSBuild" tab:

    MSB3073: The command "dotnet --list-sdks" exited with code 127. 
    /Users/theome/.nuget/packages/flatsharp.compiler/7.7.0/build/FlatSharp.Compiler.targets(98,5)

This is an excerpt from the `msbuild_log.txt`:

    Project "nuget-example.csproj" (default targets):
      Exec: dotnet --list-sdks
      Exec: /var/folders/w_/mdv4sslx3sg2_xk497c1vz5h0000gn/T/MSBuildTemptheome/tmpb99fda8e72d64fe99d946dee8fd19f35.exec.cmd: line 2: dotnet: command not found
      /Users/theome/.nuget/packages/flatsharp.compiler/7.7.0/build/FlatSharp.Compiler.targets(98,5): error MSB3073: The command "dotnet --list-sdks" exited with code 127. [/Users/theome/nuget-example/nuget-example.csproj]
    Done building project "nuget-example.csproj" -- FAILED.

This is the command in the file `FlatSharp.Compiler.targets` that tries to run `dotnet`:

    <Exec Command="dotnet --list-sdks" ConsoleToMsBuild="false">
      <Output TaskParameter="ConsoleOutput" PropertyName="StdOut" />
    </Exec>

If I modify the file and add the following command:

    <Exec Command="echo $PATH" ConsoleToMsBuild="false">
      <Output TaskParameter="ConsoleOutput" PropertyName="StdOut" />
    </Exec>

then I can see in `msbuild_log.txt` that `PATH` doesn't contain the required `dotnet` path:

    Project "nuget-example.csproj" (default targets):
      Exec: echo $PATH
      Exec: /usr/bin:/bin:/usr/sbin:/sbin

- `dotnet` is installed correctly on my machine, and I can build and run the Godot project if I don't add the dependency.
- If I build and run the project from VS Code, the build process succeeds and the game launches.
- If I open Godot from a shell with `open /Applications/Godot_mono.app/` and build and run the project from Godot, it also works.
- Setting an explicit `dotnet` path in Editor Settings at `dotnet/editor/custom_exec_path` does not solve the issue.

I would expect that the project builds without errors when I launch Godot from the Finder and then use the "Build Project" option in Godot. Would it be possible to add an Editor Setting to explicitly set environment variables for the MSBuild process?

### Minimal reproduction project (MRP)

[nuget-example.zip](https://github.com/user-attachments/files/16886607/nuget-example.zip)

--- a/core/os/os.h
+++ b/core/os/os.h
@@ -205,6 +205,7 @@ public:
 	virtual String get_environment(const String &p_var) const = 0;
 	virtual void set_environment(const String &p_var, const String &p_value) const = 0;
 	virtual void unset_environment(const String &p_var) const = 0;
+	virtual void load_shell_environment() const {}
 
 	virtual String get_name() const = 0;
 	virtual String get_identifier() const;

--- a/doc/classes/ProjectSettings.xml
+++ b/doc/classes/ProjectSettings.xml
@@ -373,6 +373,10 @@
 			Forces a [i]constant[/i] delay between frames in the main loop (in milliseconds). In most situations, [member application/run/max_fps] should be preferred as an FPS limiter as it's more precise.
 			This setting can be overridden using the [code]--frame-delay &lt;ms;&gt;[/code] command line argument.
 		</member>
+		<member name="application/run/load_shell_environment" type="bool" setter="" getter="" default="false">
+			If [code]true[/code], loads the default shell and copies environment variables set by the shell startup scripts to the app environment.
+			[b]Note:[/b] This setting is implemented on macOS for non-sandboxed applications only.
+		</member>
 		<member name="application/run/low_processor_mode" type="bool" setter="" getter="" default="false">
 			If [code]true[/code], enables low-processor usage mode. When enabled, the engine takes longer to redraw, but only redraws the screen if necessary. This may lower power consumption, and is intended for editors or mobile applications. For most games, because the screen needs to be redrawn every frame, it is recommended to keep this setting disabled.
 		</member>

--- a/main/main.cpp
+++ b/main/main.cpp
@@ -1037,6 +1037,7 @@ Error Main::setup(const char *execpath, int argc, char *argv[], bool p_second_ph
 
 	Vector<String> breakpoints;
 	bool delta_smoothing_override = false;
+	bool load_shell_env = false;
 
 	String default_renderer = "";
 	String default_renderer_mobile = "";
@@ -2642,12 +2643,18 @@ Error Main::setup(const char *execpath, int argc, char *argv[], bool p_second_ph
 	OS::get_singleton()->_allow_hidpi = GLOBAL_DEF("display/window/dpi/allow_hidpi", true);
 	OS::get_singleton()->_allow_layered = GLOBAL_DEF_RST("display/window/per_pixel_transparency/allowed", false);
 
+	load_shell_env = GLOBAL_DEF("application/run/load_shell_environment", false);
+
 #ifdef TOOLS_ENABLED
 	if (editor || project_manager) {
 		// The editor and project manager always detect and use hiDPI if needed.
 		OS::get_singleton()->_allow_hidpi = true;
+		load_shell_env = true;
 	}
 #endif
+	if (load_shell_env) {
+		OS::get_singleton()->load_shell_environment();
+	}
 
 	if (separate_thread_render == -1) {
 		separate_thread_render = (int)GLOBAL_DEF("rendering/driver/threads/thread_model", OS::RENDER_THREAD_SAFE) == OS::RENDER_SEPARATE_THREAD;

--- a/platform/macos/os_macos.h
+++ b/platform/macos/os_macos.h
@@ -78,6 +78,8 @@ public:
 	virtual void set_cmdline_platform_args(const List<String> &p_args);
 	virtual List<String> get_cmdline_platform_args() const override;
 
+	virtual void load_shell_environment() const override;
+
 	virtual String get_name() const override;
 	virtual String get_distribution_name() const override;
 	virtual String get_version() const override;

--- a/platform/macos/os_macos.mm
+++ b/platform/macos/os_macos.mm
@@ -238,6 +238,32 @@ List<String> OS_MacOS::get_cmdline_platform_args() const {
 	return launch_service_args;
 }
 
+void OS_MacOS::load_shell_environment() const {
+	static bool shell_env_loaded = false;
+	if (unlikely(!shell_env_loaded)) {
+		shell_env_loaded = true;
+		if (OS::get_singleton()->has_environment("TERM") || OS::get_singleton()->has_environment("__GODOT_SHELL_ENV_SET")) {
+			return; // Already started from terminal, or other the instance with the shell environment, do nothing.
+		}
+		String pipe;
+		List<String> args;
+		args.push_back("-c");
+		args.push_back(". /etc/zshrc;. /etc/zprofile;. ~/.zshenv;. ~/.zshrc;. ~/.zprofile;env");
+		Error err = OS::get_singleton()->execute("zsh", args, &pipe);
+		if (err == OK) {
+			Vector<String> env_vars = pipe.split("\n");
+			for (const String &E : env_vars) {
+				Vector<String> tags = E.split("=", 2);
+				if (tags.size() != 2 || tags[0] == "SHELL" || tags[0] == "USER" || tags[0] == "COMMAND_MODE" || tags[0] == "TMPDIR" || tags[0] == "TERM_SESSION_ID" || tags[0] == "PWD" || tags[0] == "OLDPWD" || tags[0] == "SHLVL" || tags[0] == "HOME" || tags[0] == "DISPLAY" || tags[0] == "LOGNAME" || tags[0] == "TERM" || tags[0] == "COLORTERM" || tags[0] == "_" || tags[0].begins_with("__CF") || tags[0].begins_with("XPC_") || tags[0].begins_with("__GODOT")) {
+					continue;
+				}
+				OS::get_singleton()->set_environment(tags[0], tags[1]);
+			}
+		}
+		OS::get_singleton()->set_environment("__GODOT_SHELL_ENV_SET", "1");
+	}
+}
+
 String OS_MacOS::get_name() const {
 	return "macOS";
 }

